File: src/R/sum_reads.R
#!/usr/bin/Rscript
# Generates matrix containing read frequency of all barcode combinations

suppressMessages(library(tidyr))
suppressMessages(library(dplyr))

args <- commandArgs()

read_freq_file <- args[6] # read frequency file (frequency, barcodes)
output_file <- args[7] # output filename (R1, R2, R3, PKR, fragments)
observed_barcodes_file <- args[8] # barcode combination file (sorted barcodes w no dups)

# Read in barcodes file
observed_barcodes <- read.csv(observed_barcodes_file, header=F, sep="")
# Get R1, R2, R3 vectors
R1_barcodes <- unlist(lapply(observed_barcodes$V1,function(x){strsplit(x,",")[[1]][1]}))
R2_barcodes <- unlist(lapply(observed_barcodes$V1,function(x){strsplit(x,",")[[1]][2]}))
R3_barcodes <- unlist(lapply(observed_barcodes$V1,function(x){strsplit(x,",")[[1]][3]}))

# Read in read frequency file
read_freq <- read.csv(read_freq_file, header=F, sep="")
colnames(read_freq) <- c("freq", "barcode")
# Get PKR vector
PKRs <- unlist(lapply(read_freq$barcode, function(x){strsplit(x,",")[[1]][4]}))

# Generate dataframe containing all combinations of R1, R2, R3, and PKR
combos <- tidyr::crossing(R1=R1_barcodes, R2=R2_barcodes, R3=R3_barcodes, PKR=PKRs)
# Drop PKR column if NA
if(is.na(combos$PKR[1])){
  combos <- dplyr::select(combos,-PKR)
}
# Concatenate elements to get barcode
combos$barcode <- apply(combos, 1, paste, collapse=",")

# Match possible combinations to observed combinations and get read frequency of matches
combos$fragments <- read_freq$freq[match(combos$barcode, read_freq$barcode)]
combos[is.na(combos)] <- 0
combos <- dplyr::select(combos,-barcode)

# Write output file
write.csv(combos, output_file, quote=F, row.names=F)

File: src/python/sum_reads.py
#!/usr/bin/env python3

"""
Generates a csv containing read frequencies of all possible barcode combinations
"""

import argparse

parser = argparse.ArgumentParser(description="Generates a csv containing read frequencies of all possible barcode combinations")
parser.add_argument("-b", "--barcode_file", help="File containing unique observed barcodes")
parser.add_argument("-r", "--read_freq_file", help="File containing read frequencies of observed barcodes")
parser.add_argument("-o", "--output_file", help="Name of output file")

args = parser.parse_args()

if getattr(args, "barcode_file") is None:
    print("ERROR: Barcode file not provided\n")
    parser.parse_args(["-h"])
    
if getattr(args, "read_freq_file") is None:
    print("ERROR: Read frequency file not provided\n")
    parser.parse_args(["-h"])
    
barcode_file = getattr(args, "barcode_file")
read_freq_file = getattr(args, "read_freq_file")
output_file = getattr(args, "output_file")
    
# use generator to split lines rather than reading into memory
def get_split_lines(file_name, delimiter):
    with open(file_name, "r") as f:
        for line in f:
            yield line.rstrip().split(sep=delimiter)

# create list of all possible barcode combinations
def get_barcode_combos(barcode_file):
    unique_R1 = list({line[0] for line in get_split_lines(barcode_file, delimiter=",")})
    unique_R2 = list({line[1] for line in get_split_lines(barcode_file, delimiter=",")})
    unique_R3 = list({line[2] for line in get_split_lines(barcode_file, delimiter=",")})
    # get unique PKR values only if PKR field exists in barcode file
    unique_PKR = list({line[3] for line in get_split_lines(barcode_file, ",") if len(line) > 3})
    
    if unique_PKR:
        possible_barcodes = [",".join([R1,R2,R3,PKR]) for R1 in unique_R1 for R2 in unique_R2 for R3 in unique_R3 for PKR in unique_PKR]
    else:
        possible_barcodes = [",".join([R1,R2,R3]) for R1 in unique_R1 for R2 in unique_R2 for R3 in unique_R3]
        
    return possible_barcodes    

# create dictionary of read frequencies of all possible barcode combinations               
def get_read_freqs(barcode_file, read_freq_file):
    # make dictionary of observed barcodes and associated read counts
    read_freq_dict = {line[1]:int(line[0]) for line in get_split_lines(read_freq_file, delimiter="\t")}
    
    # initialize dictionary with read counts of 0 for each possible barcode combination
    init_dict = {barcode:0 for barcode in get_barcode_combos(barcode_file)}
    
    # update with counts of observed barcodes
    init_dict.update(read_freq_dict)
    
    return init_dict

read_freqs = get_read_freqs(barcode_file, read_freq_file)

print(f"Writing table of barcode read counts to {output_file}\n")

# write dictionary to csv file    
with open((output_file), "w") as f:
    # write header based on presence/absence of PKR field
    if len(list(read_freqs.keys())[0].split(",")) == 4:        
        f.write("R1,R2,R3,PKR,fragments\n")
    else:
        f.write("R1,R2,R3,fragments\n")
        
    for key in read_freqs.keys():
        f.write("%s,%d\n" % (key, read_freqs[key]))

print("Finished writing csv file\n")